tenants-certified-attributes-updater/src/main/scala/it/pagopa/interop/tenantscertifiedattributesupdater/util/Utils.scala

```diff
@@ -159,6 +159,27 @@ object Utils {
       }
   }
 
+  def processRevocations(
+    revoker: (PersistentExternalId, AttributeInfo) => Future[Unit],
+    list: List[(PersistentExternalId, List[AttributeInfo])]
+  )(implicit
+    actorSystem: ActorSystem[_],
+    contexts: Seq[(String, String)],
+    executionContext: ExecutionContext,
+    logger: LoggerTakingImplicit[ContextFieldsToLog]
+  ): Future[Unit] = list match {
+    case Nil       => Future.unit
+    case xs :: Nil => {
+      logger.info(s"Processing tenant ${xs._1} with attributes ${xs._2}")
+      Future.traverse(xs._2)(revoker(xs._1, _)).void
+    }
+    case xs :: xss =>
+      logger.info(s"Processing tenant ${xs._1} with attributes ${xs._2}")
+      Future.traverse(xs._2)(revoker(xs._1, _)).flatMap { _ =>
+        processRevocations(revoker, xss)
+      }
```

Comment:
can you avoid using `._1` and `._2`, since we're pattern matching you can use
`case (externalId, attributeInfos) :: Nil` or `case (externalId, attributeInfos) :: xss`

Also, doesn't the whole recursive function gets substituted by a `_.map` or a `Future.traverse`? Something like

```scala
  def processRevocations(
    revoker: (PersistentExternalId, AttributeInfo) => Future[Unit],
    list: List[(PersistentExternalId, List[AttributeInfo])]
  )(implicit
    actorSystem: ActorSystem[_],
    contexts: Seq[(String, String)],
    executionContext: ExecutionContext,
    logger: LoggerTakingImplicit[ContextFieldsToLog]
  ): Future[Unit] = 
    Future.traverse(list){ case (externalId, attributeInfos) => 
      logger.info(s"Processing tenant ${externalId} with attributes ${attributeInfos.mkString(",")}")
      Future.traverse(attributeInfos)(revoker(externalId, _))
    }.map(_ => ()) //Don't use .void, it's from cats
```

Reply:
btw you're getting better at writing `Future.flatmap` recursive functions, but let's leverage them only when it's strictly needed :)

Reply:
`Main.scala` you mean? :D I added in commons some utility functions that are `Future` compatible. You may want to use `Future.traverseWithLatch(n)` that automatically groups the inputs.